File: tmb/archive/snow.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(year_n)
  DATA_INTEGER(size_n)
  DATA_MATRIX(imm_n_at_size_obs)
  DATA_MATRIX(mat_n_at_size_obs)
  DATA_MATRIX(prob_term_molt)
  DATA_MATRIX(size_trans)
  DATA_SCALAR(log_mu_m)
  DATA_SCALAR(sigma_numbers_imm)
  DATA_SCALAR(sigma_numbers_mat)

  PARAMETER_VECTOR(log_n_imm);
  PARAMETER_VECTOR(log_n_mat);
  PARAMETER_MATRIX(nat_m);
  PARAMETER_VECTOR(log_recruits);
  PARAMETER(sigma_m);
  PARAMETER(prop_rec);
  
  // try this with natural mortality for immature and mature animals split...molting could be an issue
  matrix<Type> imm_n_at_size_pred(year_n+1,size_n);
  matrix<Type> mat_n_at_size_pred(year_n+1,size_n);
  vector<Type> temp_imm(size_n);
  vector<Type> temp_mat(size_n);
  vector<Type> trans_imm(size_n);
  
 
  Type imm_like;
  Type mat_like;
  Type nat_m_like;
  Type obj_fun;

  // End of specifications section
  // =============================

  // initial year numbers at size
  for (int size=0;size<size_n;size++) 
  {
	imm_n_at_size_pred(0,size) = exp(log_n_imm(size));
	mat_n_at_size_pred(0,size) = exp(log_n_mat(size));
  }
  
  // Project each numbers at size matrix
   for (int year=0;year<year_n;year++) 
   {
	  // molting, growth, and mortality (not sure why this didn't work...)
      // imm_n_at_size_pred.row(year+1) = imm_n_at_size_pred.row(year) * (1-prob_term_molt.row(year)) * exp(-1*nat_m.row(year)) * size_trans;
      // mat_n_at_size_pred.row(year+1) = (imm_n_at_size_pred.row(year) * (prob_term_molt.row(year)) * exp(-1*nat_m.row(year)) * size_trans) + mat_n_at_size_pred.row(year) * exp(-1*nat_m.row(year));

     // natural mortality
      for (int size=0;size<size_n;size++) 
	   {
		temp_imm(size) = imm_n_at_size_pred(year,size) * exp(-1*exp(nat_m(year,size)));
		temp_mat(size) = mat_n_at_size_pred(year,size) * exp(-1*exp(nat_m(year,size)));
	   }	
	  
	  // growth...why doesn't the first one worK?
	   trans_imm = size_trans * temp_imm;
	  
	  // maturity
	   for (int size=0;size<size_n;size++) 
	   {
	  	 imm_n_at_size_pred(year+1,size) = trans_imm(size) * (1-prob_term_molt(year,size));	
	  	 mat_n_at_size_pred(year+1,size) = trans_imm(size) * prob_term_molt(year,size) + temp_mat(size);
       }
	  
	  // add recruitment
	   imm_n_at_size_pred(year+1,0) = exp(log_recruits(year));
   }
  
  
  // Likelihood components
  // immature numbers at size data
  imm_like = 0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
    if (imm_n_at_size_obs(year,size) >0)
     imm_like += imm_n_at_size_obs(year,size)*log(imm_n_at_size_pred(year,size)/imm_n_at_size_obs(year,size));
  imm_like = -1*imm_like;
  
  // mature numbers at size data
  mat_like = 0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
    if (mat_n_at_size_obs(year,size) >0)
     mat_like += mat_n_at_size_obs(year,size)*log(mat_n_at_size_pred(year,size)/mat_n_at_size_obs(year,size));
  mat_like = -1*mat_like;
  
  nat_m_like =0;
  for (int year=0;year<year_n;year++)
   for (int size=0;size<size_n;size++)
   nat_m_like += -1*dnorm(nat_m(year,size), log_mu_m, sigma_m, true);
  
  obj_fun = imm_like + mat_like + nat_m_like;

  REPORT(imm_n_at_size_pred);
  REPORT(mat_n_at_size_pred);
  REPORT(nat_m);
  REPORT(log_recruits);
  REPORT(imm_like);
  REPORT(mat_like);
  REPORT(nat_m_like);
  REPORT(obj_fun);


  return(obj_fun);

}

File: tmb/archive/snow.R
#==========================================================================================
#install.packages('TMB', type = 'source')
require(TMB)
library(reshape2)
library(ggplot2)
library(ggridges)
setwd("tmb/")
compile("snow.cpp") 
dyn.load(dynlib("snow"))

#gdbsource("models/snow.dll")

################################################################################
# INPUT DATA
############
year_n <- scan("dat_out.DAT",skip=2,n=1,quiet=T)
years <- scan("dat_out.DAT",skip=4,n=year_n,quiet=T)
size_n <- scan("dat_out.DAT",skip=6,n=1,quiet=T)
sizes <- scan("dat_out.DAT",skip=8,n=size_n,quiet=T)

imm_n_at_size_obs <- matrix(scan("dat_out.DAT",skip=10,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
mat_n_at_size_obs <- matrix(scan("dat_out.DAT",skip=44,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)

prob_term_molt <- matrix(scan("dat_out.DAT",skip=78,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
size_trans <- (matrix(scan("dat_out.DAT",skip=112,n=size_n*size_n,quiet=T),ncol=size_n,byrow=T))


################################################################################
# INPUT INITIAL PARAMETER VALUES
################################
log_n_imm <- scan("dat_out.PIN",skip=2,n=size_n,quiet=T)
log_n_mat <- scan("dat_out.PIN",skip=4,n=size_n,quiet=T)
nat_m <- matrix(scan("dat_out.PIN",skip=6,n=year_n*size_n,quiet=T),ncol=size_n,byrow=T)
recruits <- scan("dat_out.PIN",skip=40,n=year_n,quiet=T)
log_mu_m <- scan("dat_out.PIN",skip=42,quiet=T)
log_mu_m <- -1.2
sigma_m <- scan("dat_out.PIN",skip=44,quiet=T)
################################################################################
# Data vector
data <- list(year_n=year_n,size_n=size_n,imm_n_at_size_obs=imm_n_at_size_obs,
             mat_n_at_size_obs=mat_n_at_size_obs,prob_term_molt=prob_term_molt,
             size_trans=size_trans)

parameters <- list(log_n_imm=log_n_imm,log_n_mat=log_n_imm,
                   nat_m=nat_m,recruits=recruits,log_mu_m=log_mu_m,
                   sigma_m=sigma_m)  

parameters <- list(log_n_imm=log(imm_n_at_size_obs[1,]),log_n_mat=log(mat_n_at_size_obs[1,]),
                   nat_m=nat_m,recruits=imm_n_at_size_obs[,1],log_mu_m=log_mu_m,
                   sigma_m=sigma_m)
 write.csv(log(imm_n_at_size_obs[,1]),'rec.csv')            
# When I was testing the code
#map<-list(log_n_imm=rep(factor(NA),length(log_n_imm)),
#          log_n_mat=rep(factor(NA),length(log_n_mat)),
#          nat_m=matrix(factor(NA),ncol=ncol(nat_m),nrow=nrow(nat_m)),
#          recruits=rep(factor(NA),length(recruits)))
map<-list(nat_m=matrix(factor(NA),ncol=ncol(nat_m),nrow=nrow(nat_m)))
map<-list(sigma_m=factor(NA),log_mu_m=factor(NA))
map<-list(sigma_m=factor(NA),log_mu_m=factor(NA),
          log_n_imm=rep(factor(NA),length(log_n_imm)),
          log_n_mat=rep(factor(NA),length(log_n_imm)))
#map<-list(log_n_imm=rep(factor(NA),length(log_n_imm)),
#          log_n_mat=rep(factor(NA),length(log_n_imm)),
#          recruits=rep(factor(NA),length(recruits)))
#print(data)
#print(parameters)

################################################################################
model <- MakeADFun(data, parameters, DLL="snow",silent=T,map=map,random='nat_m')

# test code - for checking for minimization
xx <- model$fn(model$env$last.par)
#print(model$report())
#cat(model$report()$obj_fun,model$report()$Like1,model$report()$Like2,model$report()$Like3,"\n")

# Actual minimzation (with some "Bonus" parameters from nlminb)
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=100000,iter.max=1000))
fit <- nlminb(model$par, model$fn, model$gr, control=list(eval.max=1,iter.max=1))
best <- model$env$last.par.best
rep <- sdreport(model)
print(best)
print(rep)
print(model$report()$imm_like)
print(model$report()$mat_like)
print(model$report()$nat_m_like)
cat(model$report()$obj_fun,model$report()$Like1,model$report()$Like2,model$report()$Like3,"\n")
rep <- sdreport(model)
print(summary(rep))


tmp2<-(model$report()$temp_imm)
tmp3<-(model$report()$trans_imm)

out_trans<-tmp2%*%size_trans
out_trans1<-tmp2%*%t(size_trans)

plot(tmp2,ylim=c(0,max(tmp2)))
lines(c(tmp3),col=3)
lines(c(out_trans),col=2)
#=========================================================
# model fits and estimated parameters
#+========================================================
par(mfrow=c(6,6),mar=c(.1,.1,.1,.1),oma=c(4,4,1,1))
for(x in 1:nrow(imm_n_at_size_obs))
{
  plot(imm_n_at_size_obs[x,],type='l',ylim=c(0,max(imm_n_at_size_obs[x,],model$report()$imm_n_at_size_pred[x,])),
       yaxt='n',xaxt='n')
  lines(model$report()$imm_n_at_size_pred[x,],lty=2,col=2)
  legend('topright',bty='n',legend=years[x])
}

par(mfrow=c(6,6),mar=c(.1,.1,.1,.1),oma=c(4,4,1,1))
for(x in 1:nrow(imm_n_at_size_obs))
{
  plot(mat_n_at_size_obs[x,],type='l',ylim=c(0,max(mat_n_at_size_obs[x,],model$report()$mat_n_at_size_pred[x,])),
       yaxt='n',xaxt='n')
  lines(model$report()$imm_n_at_size_pred[x,],lty=2,col=2)
  legend('topright',bty='n',legend=years[x])
}

plot_m<-exp(model$report()$nat_m)
colnames(plot_m)<-sizes
rownames(plot_m)<-years
in_dat<-melt(plot_m)
colnames(in_dat)<-c("Year","Size","nat_m")

size_dat<- ggplot(dat=in_dat) 
size_dat <- size_dat + geom_density_ridges(aes(x=Size, y=Year, height = nat_m,
                                               group = Year, 
                                               fill=stat(y),alpha=.9999),stat = "identity",scale=5) +
  # scale_fill_viridis_c()+
  theme_bw() +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 0),
        axis.title.y=element_blank(),
        axis.title.x=element_blank()) 
print(size_dat)
